File: AccuPay/Domain/Calendars/PayCalendar.vb
Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema

Namespace Global.AccuPay.Entity

    <Table("calendar")>
    Public Class PayCalendar

        <Key>
        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property RowID As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property Created As Date

        Public Property CreatedBy As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Computed)>
        Public Property LastUpd As Date?

        Public Property LastUpdBy As Integer?

        Public Property Name As String

    End Class

End Namespace

File: Accupay.Data/Entities/PayCalendar.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Data.Entities
{
    [Table("calendar")]
    public class PayCalendar
    {
        public const string DefaultName = "[Default]";

        private PayCalendar()
        {
            // create a factory method to create new Daytypes
        }

        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int? RowID { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public DateTime Created { get; set; }

        public int? CreatedBy { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Computed)]
        public DateTime? LastUpd { get; set; }

        public int? LastUpdBy { get; set; }

        public string Name { get; set; }

        public bool IsDefault { get; set; }

        public static PayCalendar NewCalendar(string calendarName)
        {
            return new PayCalendar()
            {
                IsDefault = false,
                Name = calendarName?.Trim(),
            };
        }

        public static PayCalendar CreateDefaultCalendar()
        {
            return new PayCalendar()
            {
                IsDefault = true,
                Name = DefaultName,
            };
        }
    }
}
